proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1041 @@
+
+  * In all versions of GHC released before the end of 2025, the default
+    language edition will be ``GHC2021``. In addition, a new warning
+    ``-Wmissing-language-edition``, on by default, will inform users
+    that they should specify a language edition.
+
+  * Starting in 2026, the default language edition when none is specified
+    will be the latest ``Stable`` edition available. (The warning continues,
+    unchanged.)
+
+  * In GHCi, the default language edition will be the latest ``Stable``
+    edition. ``-Wmissing-language-edition`` will be off by default.
+
+- A language edition can control almost all behaviors of GHC. The meaning
+  (or existence) of other flags can depend on language edition. While
+  we will not implement it this way, we can imagine that GHC becomes
+  a set of programs that happen to share a binary; the choice of which
+  program is chosen by the language edition.
+
+  The one restriction on the expressive power of language editions
+  is that build products of different language editions must be
+  compatible. We expect the Haskell ecosystem to contain packages
+  compiled with a variety of language editions, and they must work
+  together. The word *compatible* above is doing some heavy lifting,
+  in that it contains a notion that the interface (i.e. exported
+  symbols, their types, etc.) can be translated from one language
+  edition to another. This aspect of Haskell's design has always
+  been present, in that different files can be compiled with different
+  langauge extensions. We often don't notice this, because we have
+  been careful with language extensions not to make the translation
+  apparent.
+
+  Although language editions have wide authority, we must be tasteful
+  in how they work. It would be problematic to have ``Stable2024``
+  and ``Stable2027`` disagree on the meaning of widely used features.
+  Yet I think it's best to be maximally expressive here, relying on
+  our future selves not to abuse our power.
+
+  (In the conversation, it was suggested that a ``Python`` language
+  edition would be beyond reason... but actually I think such a thing
+  would be lovely, if the necessary work was done to translate the
+  interfaces between Python and Haskell.)
+
+- For backward compatibility, a language edition can be specified
+  at an arbitrary place in a command-line invocation of GHC, or in a
+  ``LANGUAGE`` pragma in a file. Even though it might come later in
+  a command line, the edition can affect the meaning of command-line
+  arguments that precede it.
+
+- A cabal file will allow a new
+  field ``language-edition``, available both at top-level and in
+  build-product stanzas. This will specify the language edition. To
+  support backward compatibility, this will use the ``default-language``
+  setting if that is available, and omitting the ``language-edition`` will
+  use the default. At some point, it is expected that ``language-edition``
+  will become required.
```

Comment:
Minor point, of course, but why not simply keep using `default-language`?

Reply:
Because of the word "default". That word implies that this is the language for all modules except those that specify some other language, but I think the cabal file should be authoritative about language edition. That is, I want to be able to know that a package is stable -- in the sense of https://github.com/ghc-proposals/ghc-proposals/pull/625 -- just by looking at the cabal file. Maybe this is problematic if someone really wants a package mixing language editions? I suppose if that is a use case we want to support, there could be `language-edition: as-specified-in-each-module` or something. Or maybe this field is not required. But I'm pretty sure I don't want it to just specify a default.

Reply:
Ah, I get it; this would _not_ be overridable from the individual modules then